mypyc/genops.py

```diff
@@ -1958,27 +1970,36 @@ def add_call_to_callable_class(self, call_fn: FuncIR) -> None:
 
     def instantiate_callable_class(self) -> Value:
         """Assigns a callable class to a register named after the given function definition."""
-        fdef = self.fn_info.fdef
+        fitem = self.fn_info.fitem
 
         fullname = '{}.{}'.format(self.module_name, self.fn_info.callable_class.name)
-        func_reg = self.add(Call(RInstance(self.fn_info.callable_class), fullname, [], fdef.line))
+        func_reg = self.add(Call(RInstance(self.fn_info.callable_class), fullname, [],
+                                 fitem.line))
 
         # Set the callable class' environment attribute to point at the environment class
         # defined in the callable class' immediate outer scope.
-        self.add(SetAttr(func_reg, ENV_ATTR_NAME, self.fn_infos[-2].env_reg, fdef.line))
+        self.add(SetAttr(func_reg, ENV_ATTR_NAME, self.fn_infos[-2].env_reg, fitem.line))
+
+        # If we have a lambda expression, then just return the register where it is stored.
+        if isinstance(fitem, LambdaExpr):
+            self.fn_info.callable_reg = func_reg
+            return self.fn_info.callable_reg
 
-        if fdef.original_def:
+        # If it is not a lambda expression, then it must be a function definition.
+        assert isinstance(fitem, FuncDef)
```

Comment:
I'd say avoid the casing on the function type here by having `instantiate_callable_class` always return `func_reg` (without worrying about assigning it to the variable), and handle assigning that to the appropriate register for FuncDefs in `visit_func_def` (or a helper function called from there).

That will let us only need to case on whether a function is a lambda or a funcdef once. (Implicitly, as part of the visitor.)

Reply:
Done.